### PR TITLE
[Merged by Bors] - Use canonical ed25519 for ballots

### PR DESCRIPTION
**Pull request by fasmat**
## Motivation
Part of #4139, merge after https://github.com/spacemeshos/go-spacemesh/pull/4230

## Changes
- Add `SmesherID` to messages signed for the `BALLOT` domain and verify those signatures instead of extracting the PublicKey from the signature

## Test Plan
- existing tests pass

## TODO

- [x] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [x] Update documentation as needed

## DevOps Notes

- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

### CODE REVIEW

File: common/types/proposal.go [hunk 1]
```diff
@@ -43,7 +43,9 @@ func (id *ProposalID) DecodeScale(d *scale.Decoder) (int, error) {
 type Proposal struct {
 	// the content proposal for a given layer and the votes on the mesh history
 	InnerProposal
-	// smesher's signature on InnerProposal
+	// the public key of the smesher that produced this proposal
+	SmesherID NodeID
```
Comment (countvonzero) [inline common/types/proposal.go]: no need for this. Proposal is always accompanied by a Ballot, which will contain NodeID

File: common/types/ballot.go [hunk 2]
```diff
@@ -44,6 +44,8 @@ type Ballot struct {
 	InnerBallot
 	// smeshers signature on InnerBallot
 	Signature EdSignature
```
Comment (countvonzero) [inline common/types/ballot.go:46]: exclude signature from the ID calculation.

Comment (fasmat) [reply common/types/ballot.go:46]: I updated the code accordingly but one test (in the `datastore` package) keeps failing and I haven't figured out why yet 🙁

Comment (fasmat) [reply common/types/ballot.go:46]: Hope it is fixed now 😅. Separating `BallotMetadata` from ballot will be done in a future PR.

**Review by countvonzero — APPROVED** [review]

**Comment by fasmat** [comment]
bors merge

File: miner/proposal_builder.go [hunk 3]
```diff
@@ -259,7 +259,7 @@ func (pb *ProposalBuilder) createProposal(
 	}
 	p.Ballot.Signature = pb.signer.Sign(signing.BALLOT, p.Ballot.SignedBytes())
 	p.Signature = pb.signer.Sign(signing.BALLOT, p.SignedBytes())
-	p.SetSmesherID(pb.signer.NodeID())
+	p.SmesherID = pb.signer.NodeID()
```
Comment (countvonzero) [inline miner/proposal_builder.go:261]: you need to move this up before signing the proposal.

Comment (countvonzero) [reply miner/proposal_builder.go:261]: unfortunately, putting SmesherID inside Ballot meaning it's part of the proposal signature.

Comment (countvonzero) [reply miner/proposal_builder.go:261]: why doesn't unittest catch this error?

Comment (fasmat) [reply miner/proposal_builder.go:261]: Fixed and updated unit tests to detect if the signature of a proposal cannot be verified

**Comment by fasmat** [comment]
bors merge

**Pull request closed without merging**